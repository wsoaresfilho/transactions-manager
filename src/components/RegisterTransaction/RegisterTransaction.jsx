import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import './styles/register.css';

const transactionType = {
    CREDIT: 'credit',
    DEBIT: 'debit',
};

const defaultState = {
    isCredit: true,
    description: '',
    value: 0.0,
};

class RegisterTransaction extends PureComponent {
    constructor(props) {
        super(props);

        this.formRef = React.createRef();
        this.toggleRef = React.createRef();
        this.state = defaultState;

        this.onSubmit = this.onSubmit.bind(this);
        this.handleTransactionTypeChange = this.handleTransactionTypeChange.bind(
            this
        );
        this.handleTransitionEnd = this.handleTransitionEnd.bind(this);
        this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
        this.handleMoneyChange = this.handleMoneyChange.bind(this);
    }

    handleTransitionEnd() {
        const signalRotation = this.state.isCredit ? '' : '-';
        this.toggleRef.current.style.transform = `rotateY(${signalRotation}15deg)`;
        setTimeout(() => (this.toggleRef.current.style.transform = ''), 400);
    }

    getTypeText() {
        return this.state.isCredit
            ? transactionType.CREDIT
            : transactionType.DEBIT;
    }

    onSubmit(event) {
        event.preventDefault();
        const { isCredit, ...transaction } = this.state;
        const value = this.state.isCredit
            ? transaction.value
            : transaction.value * -1;
        this.props.addTransaction({
            ...transaction,
            value,
        });
        this.setState(defaultState);
    }

    handleDescriptionChange(event) {
        this.setState({
            description: event.target.value,
        });
    }

    handleMoneyChange(event) {
        const moneyValue = parseFloat(event.target.value);

        this.setState({
            value: moneyValue,
        });
    }

    handleTransactionTypeChange(event) {
        const tranType = event.target.value;
        const isCredit = tranType === transactionType.CREDIT;
        this.setState({
            isCredit,
        });
    }

    render() {
        const { isCredit, description, value } = this.state;
        const typeText = this.getTypeText();
        const typeClassname = classNames({
            credit: isCredit,
            debit: !isCredit,
            'register-transaction__radio': true,
        });
        return (
            <div className='register-transaction__container'>
                <h1 className='register-transaction__title'>
                    Register Transaction Form
                </h1>
                <form
                    className='register-transaction__form'
                    onSubmit={this.onSubmit}
                    ref={this.formRef}
                >
                    <input
                        required
                        id='description-input'
                        className='register-transaction__input'
                        type='text'
                        value={description}
                        placeholder='Transaction description'
                        onChange={this.handleDescriptionChange}
                    />
                    <input
                        required
                        id='money-input'
                        className='register-transaction__input'
                        type='number'
                        min='0.00'
                        step='0.01'
                        value={value}
                        placeholder='Transaction value'
                        onChange={this.handleMoneyChange}
                    />

                    <div className={typeClassname} ref={this.toggleRef}>
                        <input
                            onChange={this.handleTransactionTypeChange}
                            type='radio'
                            id='type-credit'
                            name='type'
                            value={transactionType.CREDIT}
                            checked={isCredit}
                        />
                        <label id='type-credit__label' htmlFor='type-credit'>
                            Credit
                        </label>

                        <input
                            onChange={this.handleTransactionTypeChange}
                            type='radio'
                            id='type-debit'
                            name='type'
                            value={transactionType.DEBIT}
                            checked={!isCredit}
                        />
                        <label id='type-debit__label' htmlFor='type-debit'>
                            Debit
                        </label>

                        <div
                            id='flap'
                            onTransitionEnd={this.handleTransitionEnd}
                        >
                            <span className='content'>{typeText}</span>
                        </div>
                    </div>

                    <button
                        className='register-transaction__button'
                        type='submit'
                    >
                        Add Transaction
                    </button>
                </form>
            </div>
        );
    }
}

RegisterTransaction.propTypes = {
    addTransaction: PropTypes.func.isRequired,
};

export default RegisterTransaction;
