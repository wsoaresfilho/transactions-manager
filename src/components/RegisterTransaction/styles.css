:root {
  --credit-color: #04da97;
  --debit-color: red;
  --border-width: 6px;
  --border-radius: 20px;
  --item-box-shadow: 1px 1px 5px 0px rgba(0, 0, 0, 0.5);
  --item-box-shadow-hover: 1px 1px 6px 1px rgba(0, 0, 0, 0.5);
  --radio-font-size: 1.2em;
}

:focus {
  outline: var(--credit-color) auto 2px;
}

.register-transaction__container {
  -webkit-box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.5);
  -moz-box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.5);
  box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.5);
  background-color: #fffff0;
  padding: 15px;
}

.register-transaction__title {
  width: 100%;
  text-align: center;
  font-size: 1.8em;
}

.register-transaction__form {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
}

.register-transaction__form > * {
  margin-bottom: 15px;
  margin-right: 20px;
}

.register-transaction__form .register-transaction__input {
  border: none;
  flex-basis: 250px;
  box-shadow: var(--item-box-shadow);
  padding: 5px 10px;
  min-height: 35px;
  flex-grow: 1;
}

.register-transaction__form .register-transaction__input:focus {
  box-shadow: var(--item-box-shadow-hover);
}

.register-transaction__form input:first-child {
  flex-grow: 3;
}

.register-transaction__radio input:checked + label {
  color: transparent;
}

.register-transaction__radio {
  position: relative;
  border-radius: var(--border-radius);
  transition: transform cubic-bezier(0, 0, 0.3, 2) 0.4s;
  transform-style: preserve-3d;
  perspective: 800px;
  min-width: 225px;
  box-shadow: var(--item-box-shadow);
}

.register-transaction__radio:focus-within {
  box-shadow: var(--item-box-shadow-hover);
}

.register-transaction__radio.credit {
  border: solid var(--border-width) var(--credit-color);
}

.register-transaction__radio.debit {
  border: solid var(--border-width) var(--debit-color);
}

.register-transaction__radio > input[type='radio'] {
  width: 1px;
  height: 1px;
  margin: -1px;
  border: 0;
  padding: 0;
  clip: rect(0 0 0 0);
  overflow: hidden;
}

.register-transaction__radio > #type-credit:checked ~ #flap {
  transform: rotateY(-180deg);
}

.register-transaction__radio > #type-credit:checked ~ #flap > .content {
  transform: rotateY(-180deg);
}

.register-transaction__radio > #type-debit:checked ~ #flap {
  transform: rotateY(0deg);
}

.register-transaction__radio > label {
  display: inline-block;
  padding: 5px 26px;
  text-align: center;
  cursor: pointer;
  font-size: var(--radio-font-size);
}

.register-transaction__radio > label,
.register-transaction__radio > #flap {
  font-weight: bold;
  text-transform: capitalize;
}

.register-transaction__radio > #flap {
  position: absolute;
  top: calc(0px - var(--border-width));
  left: 50%;
  height: calc(100% + var(--border-width) * 2);
  width: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  border-top-right-radius: var(--border-radius);
  border-bottom-right-radius: var(--border-radius);
  transform-style: preserve-3d;
  transform-origin: left;
  transition: transform cubic-bezier(0.4, 0, 0.2, 1) 0.5s;
}

.register-transaction__radio.credit > #flap {
  background-color: var(--credit-color);
}

.register-transaction__radio.debit > #flap {
  background-color: var(--debit-color);
}

.register-transaction__radio > #flap > .content {
  color: #fffff0;
  transition: transform 0s linear 0.25s;
  transform-style: preserve-3d;
  font-size: var(--radio-font-size);
}

.register-transaction__form button {
  background-color: var(--credit-color);
  box-shadow: var(--item-box-shadow);
  min-width: 200px;
  color: #fffff0;
  min-height: 45px;
  /* min-width: 235px; */
}

.hover-shadow:hover,
.hover-shadow:focus {
  box-shadow: var(--item-box-shadow-hover);
}
