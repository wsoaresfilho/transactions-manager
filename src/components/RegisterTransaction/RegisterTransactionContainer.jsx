import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import RegisterTransaction from './RegisterTransaction';
import { addTransaction as addTransactionAction } from '../../actions';

export const RegisterTransactionContainer = props => {
    const { addTransaction, className } = props;
    return (
        <RegisterTransaction
            className={className}
            addTransaction={addTransaction}
        />
    );
};

RegisterTransactionContainer.defaultProps = {
    className: null,
};

RegisterTransactionContainer.propTypes = {
    addTransaction: PropTypes.func.isRequired,
    className: PropTypes.string,
};

const mapDispatchToProps = dispatch => ({
    addTransaction: transaction => dispatch(addTransactionAction(transaction)),
});

export default connect(null, mapDispatchToProps)(RegisterTransactionContainer);
