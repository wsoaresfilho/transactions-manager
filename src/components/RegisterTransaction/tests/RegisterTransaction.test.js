import React from 'react';
import { shallow } from 'enzyme';
import RegisterTransaction from '../RegisterTransaction';

describe('RegisterTransaction component', () => {
    let wrapper;
    let instance;
    let spy;
    const props = {
        addTransaction: jest.fn(),
    };

    beforeEach(() => {
        wrapper = shallow(
            <RegisterTransaction addTransaction={props.addTransaction} />
        );
        instance = wrapper.instance();
        spy = {
            description: jest.spyOn(
                RegisterTransaction.prototype,
                'handleDescriptionChange'
            ),
            money: jest.spyOn(
                RegisterTransaction.prototype,
                'handleMoneyChange'
            ),
            type: jest.spyOn(
                RegisterTransaction.prototype,
                'handleTransactionTypeChange'
            ),
            onSubmit: jest.spyOn(RegisterTransaction.prototype, 'onSubmit'),
            transition: jest.spyOn(
                RegisterTransaction.prototype,
                'handleTransitionEnd'
            ),
            focus: jest.spyOn(
                RegisterTransaction.prototype,
                'focusDescriptionInput'
            ),
        };
    });

    afterEach(() => {
        spy.description.mockClear();
        spy.money.mockClear();
        spy.type.mockClear();
        spy.focus.mockClear();
    });

    describe('should render', () => {
        it('all elements properly', () => {
            expect(
                wrapper.find('.register-transaction__container').length
            ).toEqual(1);
            expect(wrapper.find('.register-transaction__title').length).toEqual(
                1
            );
            expect(wrapper.find('.register-transaction__title').text()).toEqual(
                'Register Transaction Form'
            );
            expect(wrapper.exists('.register-transaction__form')).toBeTruthy();
            expect(
                wrapper.exists('.register-transaction__button')
            ).toBeTruthy();
        });

        it('snapshot correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when mounts', () => {
        it('shold focus on description', () => {
            expect(spy.focus).toHaveBeenCalled();
        });
    });

    describe('when input description', () => {
        it('should set description state', () => {
            const descriptionInput = wrapper.find('#description-input');
            const descriptionText = 'Testing description';
            const event = {
                target: {
                    value: descriptionText,
                },
            };
            descriptionInput.simulate('change', event);

            expect(instance.state.description).toEqual(descriptionText);
            expect(spy.description).toHaveBeenCalled();
        });
    });

    describe('when input money', () => {
        it('should set value state', () => {
            const moneyInput = wrapper.find('#money-input');
            const moneyText = '123.40';
            const event = {
                target: {
                    value: moneyText,
                },
            };
            moneyInput.simulate('change', event);

            expect(instance.state.value).toEqual(123.4);
            expect(spy.money).toHaveBeenCalled();
        });
    });

    describe('when change transaction type', () => {
        it('to debit, it should set isCredit state to false', () => {
            const debitInput = wrapper.find('#type-debit');
            const debitValue = 'debit';
            const event = {
                target: {
                    value: debitValue,
                },
            };
            debitInput.simulate('change', event);

            expect(instance.state.isCredit).toBe(false);
            expect(spy.type).toHaveBeenCalledTimes(1);
        });

        it('to credit, it should set isCredit state to true', () => {
            const creditInput = wrapper.find('#type-credit');
            const event = {
                target: {
                    value: 'credit',
                },
            };
            creditInput.simulate('change', event);

            expect(instance.state.isCredit).toBe(true);
            expect(spy.type).toHaveBeenCalledTimes(1);
        });

        it('should have a transition on flap element', () => {
            jest.useFakeTimers();
            instance.toggleRef.current = document.createElement('div');
            const flap = wrapper.find('#flap');
            flap.simulate('transitionend');

            expect(spy.transition).toHaveBeenCalled();
            expect(setTimeout).toHaveBeenCalledTimes(1);
            expect(setTimeout).toHaveBeenLastCalledWith(
                expect.any(Function),
                400
            );
        });
    });

    describe('when submit a transaction', () => {
        it('should call addTransaction', () => {
            const descriptionInput = wrapper.find('#description-input');
            const descriptionText = 'Testing description when submiting';
            const moneyInput = wrapper.find('#money-input');
            const moneyText = '333.55';
            const debitInput = wrapper.find('#type-debit');
            const debitValue = 'debit';
            const form = wrapper.find('.register-transaction__form');
            const getEvent = value => ({
                target: {
                    value,
                },
                preventDefault: jest.fn(),
            });

            descriptionInput.simulate('change', getEvent(descriptionText));
            moneyInput.simulate('change', getEvent(moneyText));
            debitInput.simulate('change', getEvent(debitValue));

            expect(instance.state).toStrictEqual({
                isCredit: false,
                description: descriptionText,
                value: parseFloat(moneyText),
            });

            form.simulate('submit', getEvent());
            expect(instance.state).toStrictEqual({
                isCredit: true,
                description: '',
                value: '',
            });
            expect(spy.onSubmit).toHaveBeenCalledTimes(1);
            expect(spy.focus).toHaveBeenCalled();
            expect(props.addTransaction).toHaveBeenCalledTimes(1);
        });
    });
});
