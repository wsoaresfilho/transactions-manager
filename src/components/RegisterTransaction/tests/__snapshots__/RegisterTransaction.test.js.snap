// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterTransaction component should render snapshot correctly 1`] = `ShallowWrapper {}`;
